### PR TITLE
New logm implementation

### LINKED ISSUE #21179: logm issue
As reported by @iamnapo in https://github.com/JuliaLang/julia/issues/21143#issuecomment-289306668

```julia
julia> B = [3 2; -5 -3]
2×2 Array{Int64,2}:
  3   2
 -5  -3

julia> expm(logm(B))
2×2 Array{Float64,2}:
 -5.33599  -5.66504
 17.5616   18.4572
```

### PR DESCRIPTION
**Pull request by iamnapo**
Fixed 21179

Original algorithm can be found [here](https://www.mathworks.com/matlabcentral/fileexchange/33393-matrix-logarithm?focused=5200819&tab=function).

### CODE REVIEW

File: base/linalg/triangular.jl [hunk 1]
```diff
@@ -1742,187 +1742,183 @@ powm(A::LowerTriangular, p::Real) = powm(A.', p::Real).'
...
-        if s > s0
-            d3 = cbrt(norm(AmI^3, 1))
-        end
-        d4 = norm(AmI^4, 1)^(1/4)
-        alpha3 = max(d3, d4)
-        if alpha3 <= theta[tmax]
-            for j = 3:tmax
-                if alpha3 <= theta[j]
-                    break
+        m = 0; m += (alpha2 <= theta[1] ? 1 : 2)
```
Comment (tkelman) [inline base/linalg/triangular.jl]: why the `= 0` then immediately `+=` ?

File: base/linalg/triangular.jl [hunk 2]
```diff
@@ -1741,188 +1741,187 @@ powm(A::LowerTriangular, p::Real) = powm(A.', p::Real).'
...
 end
 logm(A::LowerTriangular) = logm(A.').'
 
 # Auxiliary functions for logm and matrix power
 
+# Matrix square root by product form of Denman-Beavers iteration.
+#   computes the principal square root X
+#   of the matrix A using the product form of the Denman-Beavers
+#   iteration. The matrix M tends to I.
+#       Copyright (c) 2008 by N. J. Higham
```
Comment (tkelman) [inline base/linalg/triangular.jl]: he's been very nice before about granting permission to relicense derived translations of his implementations, did a previous instance of that also cover this bit of code?

File: base/linalg/triangular.jl (hunk 1 above)
Comment (iamnapo) [inline base/linalg/triangular.jl]: my mistake

File: base/linalg/symmetric.jl [hunk 3]
```diff
@@ -473,7 +473,7 @@ for (funm, func) in ([:logm,:log], [:sqrtm,:sqrt])
     @eval begin
         function ($funm){T<:Real}(A::Symmetric{T})
             F = eigfact(A)
-            if isposdef(F)
+            if all(λ -> λ ≥ 0, F.values)
```
Comment (fredrikekre) [inline base/linalg/symmetric.jl]: Why this change? `isposdef(F)` is more clear IMO and [does the same thing](https://github.com/JuliaLang/julia/blob/7eefa45bdedd81da4742c5a1649b241c09fd6424/base/linalg/eigen.jl#L30)? Same just below too.

**Comment by tkelman** [comment]
Would be good to add the cases from #21179 as tests, and maybe a handful more that are wrong or broken on master but fixed here.

File: base/linalg/triangular.jl (hunk 2 above)
Comment (iamnapo) [inline base/linalg/triangular.jl]: Well`sqrtm_dbp` is internally called by `logm`, for which he has given permission to use. So yes, I guess.

cc @higham

File: base/linalg/dense.jl [hunk 4]
```diff
@@ -551,40 +551,26 @@ julia> logm(A)
  0.0  1.0
 ```
 """
-function logm(A::StridedMatrix{T}) where T
+function logm{T}(A::StridedMatrix{T})
```
Comment (tkelman) [inline base/linalg/dense.jl]: this should go back the way it was

File: test/linalg/dense.jl [hunk 5]
```diff
@@ -454,31 +454,55 @@ end
         A5  = convert(Matrix{elty}, [1 1 0 1; 0 1 1 0; 0 0 1 1; 1 0 0 1])
         @test expm(logm(A5)) ≈ A5
 
-        A6  = convert(Matrix{elty}, [-5 2 0 0 ; 1/2 -7 3 0; 0 1/3 -9 4; 0 0 1/4 -11])
-        @test expm(logm(A6)) ≈ A6
```
Comment (tkelman) [inline test/linalg/dense.jl]: why remove this one?
(likewise with `@test (Ab^im)^(-im) ≈ Ab`)

File: base/linalg/dense.jl (hunk 4 above)
Comment (iamnapo) [inline base/linalg/dense.jl]: For future reference, is there something wrong with this, or is it just the preferred syntax?

File: test/linalg/dense.jl (hunk 5 above)
Comment (iamnapo) [inline test/linalg/dense.jl]: Because A6, Ab have negative eigenvalues so logm is not unique and the logarithm returned on these two occasions isn't the one that `expm(logm(A)) == A`.
(`Ab^im = expm(im*logm(A))`)

File: base/linalg/dense.jl (hunk 4 above)
Comment (tkelman) [inline base/linalg/dense.jl]: the `logm{T}` version is likely to be deprecated for 0.7/1.0

File: test/linalg/dense.jl (hunk 5 above)
Comment (tkelman) [inline test/linalg/dense.jl]: so the new algorithm returns a different root? should we test the new result instead?

Comment (iamnapo) [inline test/linalg/dense.jl]: I'm not sure how to test that

Comment (tkelman) [inline test/linalg/dense.jl]: compare the result against its expected value written out? is the root this algorithm now returns consistent with what any other libraries return, like wolfram or matlab or octave, or would we be making a new outlier choice?

Comment (GunnarFarneback) [inline test/linalg/dense.jl]: Now I'm not following. Even if `logm(A)` isn't unique, any such solution should satisfy `expm(logm(A)) == A`, shouldn't it? How is it a valid `logm(A)` otherwise?

Comment (iamnapo) [inline test/linalg/dense.jl]: Yes yes, my previous comment was wrong. If A has negative eigenvalues, logm is not unique and this algorithm's results may be inaccurate.

I'll work on it more.

**Comment by iamnapo** [comment]
It appears that the previous implementation was better and the problem was elsewhere.
Closing this PR and open the correct one shortly.

**Comment by YingboMa** [comment]
@iamnapo If you are using MATLAB as a comparison, be aware that [this](https://www.mathworks.com/matlabcentral/fileexchange/33393-matrix-logarithm?focused=5200819&tab=function) algorithm is not as robust as the MATLAB build-in algorithm for non-principal matrix.

**Comment by iamnapo** [comment]
Yeap, that's why I closed this PR.

**Comment by YingboMa** [comment]
Yeah, I tried to write a PR, and I cannot figure it out.
